File: 101-109_small_problems/easy_7/staggered_caps_pt2.rb
# Staggered Caps (Part 2)
# Modify the method from the previous exercise so it ignores non-alphabetic characters when determining whether it should uppercase or lowercase each letter. The non-alphabetic characters should still be included in the return value; they just don't count when toggling the desired case.

# Input: A string
# Output: A string
# Mental model: Given a string, create a new string with staggered capitalization.

# Example:

# staggered_case('I Love Launch School!') == 'I lOvE lAuNcH sChOoL!'
# staggered_case('ALL CAPS') == 'AlL cApS'
# staggered_case('ignore 77 the 444 numbers') == 'IgNoRe 77 ThE 444 nUmBeRs'

# Data structure

# Input and output are strings.

# Algorithm

# Return array

# Code
def staggered_case(string, start_case="uppercase")
  result = ''
  need_upper = start_case.include?("up") ? true : false
  string.chars.each do |char|
    if need_upper
      result += char.upcase
    else
      result += char.downcase
    end

    if char.match(/[^a-z]/i)
      need_upper = !!need_upper
    else
      need_upper = !need_upper
    end
  end
  result
end

p staggered_case('I Love Launch School!') == 'I lOvE lAuNcH sChOoL!'
p staggered_case('ALL CAPS') == 'AlL cApS'
p staggered_case('ignore 77 the 444 numbers') == 'IgNoRe 77 ThE 444 nUmBeRs'

# Further Exploration
# Modify this method so the caller can determine whether non-alphabetic characters should be counted when determining the upper/lowercase state. That is, you want a method that can perform the same actions that this method does, or operates like the previous version.

# Hint: Use a keyword argument.

File: exercises/js210_small_problems/string_and_text/staggeredCapsPt2.js
// Staggered Caps Part 2
// Modify the function from the previous exercise so that it ignores non-alphabetic characters when determining whether a letter should be upper or lower case. Non-alphabetic characters should still be included in the output string, but should not be counted when determining the appropriate case.

// Input: A string
// Output: A string
//
// Rules:
// - ignores non-alphabetic characters when determining a letter should be upper or lowercase
// - non alphabetic characters should still remain in the string
//
// Data Structures:
// string -> array - String
//
// Algorithm:
// 1. Create a variable to keep track of whether it's capitalized and assign it to true
// 2. downcase string
// 3. Split string into characters
// 4. Map string
// 5.   if variable is true
// 6.     Upcase current character
// 7.   else
//8.       Return current character
// 9.   Assign variable to opposite
// 10. Assign variable to opposite again if character is non alphabetic
// 9. Join string
// 10. Return string

function staggeredCase(str) {
  let isCapitalized = true;

  return str.toLowerCase().split('').map(function(char) {
    let newChar = '';

    if (isCapitalized) {
      newChar = char.toUpperCase();
    } else {
      newChar = char;
    }

    isCapitalized = !isCapitalized;

    if (/[^a-z]/i.test(char)) {
      isCapitalized = !isCapitalized;
    }

    return newChar;
  }).join('');
}
// Examples:

console.log(staggeredCase('I Love Launch School!') === "I lOvE lAuNcH sChOoL!");
console.log(staggeredCase('ALL CAPS') === "AlL cApS");
console.log(staggeredCase('ignore 77 the 444 numbers') === "IgNoRe 77 ThE 444 nUmBeRs");
